activerecord/lib/active_record/associations/collection_association.rb

```diff
@@ -455,13 +449,10 @@ def replace_on_target(record, skip_callbacks, replace:, inversing: false)
 
           yield(record) if block_given?
 
-          if !index && @replaced_or_added_targets.include?(record)
-            index = @target.index(record)
-          end
-
-          @replaced_or_added_targets << record if inversing || index || record.new_record?
+          was_implicit = @implicit_target_records&.delete(record)
+          (@implicit_target_records ||= []) << record if implicit
```

Comment:
Any reason why we initialize this variable with `nil` instead of an array if we know it needs to be an array?

Reply:
It avoids allocating the array when no records are inversed for an association.  (`implicit` here is true only when `add_record` is [called from `inversed_from`](https://github.com/rails/rails/pull/45399/files#diff-ceff30ddab4e756e3a70ece45076eb17ff2f587a068dae657d2ad3a265a3f0d6R279).)